### ISSUE TITLE
Possible bug in implicit interger to float conversion

### ISSUE THREAD

**Issue by franalbani** [KarpatkeyDAO/defi-protocols#50]
In many parts of the code, a number divided by `10**n_decimals` is returned, that may be `1` in case `decimals=False` (`n_decimals = 0`).

I feel this may bring problems:

```ipython3
In [25]: tokens = 2874698217465216421465

In [26]: tokens_w_d = tokens / 10**0

In [27]: tokens_w_d == tokens
Out[27]: False

In [28]: int(tokens_w_d) == tokens
Out[28]: False
```

In other words, the integer may be to big and be "rounded" when converted to float.

**Comment by sgzerbo** [comment]
Indeed. This is a wrong approach that was never fixed, at least not everywhere. Check the balance_of function in functions.py, we fixed it over there using the decimal package. But this bug has to be fixed in lots of other places...
![Screenshot_20230404-205045](https://user-images.githubusercontent.com/98768518/229946952-8004f6bd-6956-4002-8aa1-eff31309c9d0.png)

**Comment by franalbani** [comment]
Shouldn't the final type be `Decimal` also? That's the point of using it I believe. Mi intuition tells me the last `float()` casting is wrong.

**Comment by sgzerbo** [comment]
But other softwares using the defi-protocols (such as Uri's ETL) will be most likely expecting a float. Otherwise you force to work with the Decimal type

**Comment by franalbani** [comment]
In every place a monetary quantity is handled, I would consider "forcing users" using `Decimal` a good thing :D .

In the meantime, users can make explicit float casting in the places they need it and consider it not to be a problem.

**Comment by sanekto** [comment]
I agree with what Fran says, I would expect that a good library that handles countable items (like monetary things) to use Decimal. The casting is better to be done explicitly in the users code, so we let the user of the library to decide if she is ok loosing information or not. The Decimal module in python is very well known and the casting is just calling `int`, `float` or whatever.

**Comment by franalbani** [comment]
I already did this for `Agave` in #34 .

**Comment by sgzerbo** [comment]
You've convinced me. We should use Decimal instead of float
